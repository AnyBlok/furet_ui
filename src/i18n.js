/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import Vue from 'vue';
import VueI18n from 'vue-i18n';
Vue.use(VueI18n);


export const i18nConf = {
    locale: 'en',
    fallbackLocale: 'en',
    messages: {
        en: {
            menus: {
                close: 'Close',
                search: 'Search ...',
            },
            views: {
                unknown: {
                    title: 'The wanted view "{name}" is unknown',
                    message: "Report this message to the administrator",
                },
                common: {
                    create: 'Create',
                    save: 'Save',
                    edit: 'Edit',
                    cancel: 'Cancel',
                    delete: 'Delete',
                    close: 'Close',
                    actions: 'Actions',
                    more: 'More',
                },
                clients: {
                    login: {
                        button: 'Log in',
                    },
                },
            },
            fields: {
                common: {
                    required: 'this field is required',
                },
                json: {
                    invalid: 'Invalid JSON format',
                },
                file: {
                    upload: 'Upload',
                    download: 'Download',
                    delete: 'Delete',
                },
            },
        },
    },
};
export const i18n = new VueI18n(i18nConf);
