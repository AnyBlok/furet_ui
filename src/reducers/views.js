/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
export const defaultState = {}

export const views = (state = defaultState, view) => {
    const value = Object.assign({}, view);
    delete value.type;
    switch (view.type) {
        case 'UPDATE_VIEW':
            delete value.viewId;
            const values = {};
            values[view.viewId] = Object.assign({}, state[view.viewId], value);
            return Object.assign({}, state, values);
        case 'CLEAR_ACTION':
            return defaultState;
        default:
            return state
    }
}

export default views;
