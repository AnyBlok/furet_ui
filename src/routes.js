/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import Vue from 'vue';
import VueRouter from 'vue-router';
Vue.use(VueRouter);
import {json_post_dispatch_all} from './server-call';


const CustomeView = {
    template: '<furet-ui-custom-view v-bind:viewName="viewName" />',
    props: ['viewName'],
    beforeRouteEnter: (to, from, next) => {
        json_post_dispatch_all('/custom/view/' + to.params.viewName, {});
        next();
    },
    beforeRouteUpdate: (to, from, next) => {
        if (to.params.viewName != from.params.viewName) {
            json_post_dispatch_all('/custom/view/' + to.params.viewName, {});
        }
        next();
    },
}

const Action = {
    template: '<router-view></router-view>',
    beforeRouteEnter: (to, from, next) => {
        json_post_dispatch_all('/action/' + to.params.actionId, to.params);
        next();
    },
    beforeRouteUpdate: (to, from, next) => {
        if (to.params.actionId != from.params.actionId)
            json_post_dispatch_all('/action/' + to.params.actionId, to.params);
        next();
    },
};

const View = {
    template: `<furet-ui-view
                    v-bind:spaceId="spaceId"
                    v-bind:menuId="menuId"
                    v-bind:actionId="actionId"
                    v-bind:viewId="viewId"
                    v-bind:dataId="dataId"
                    v-bind:mode="mode"
                />`,
    props: ['spaceId', 'menuId', 'actionId', 'viewId', 'dataId', 'mode'],
    beforeRouteEnter: (to, from, next) => {
        json_post_dispatch_all('/view/' + to.params.viewId, to.params);
        next();
    },
    beforeRouteUpdate: (to, from, next) => {
        if (to.params.viewId != from.params.viewId)
            json_post_dispatch_all('/view/' + to.params.viewId, to.params);
        next();
    },
};


export const router = new VueRouter({
    routes: [
        {
            path: '/custom/view/:viewName',
            name: 'custom_view',
            props: true,
            component: CustomeView,
        },
        {
            path: '/space/:spaceId',
            name: 'space',
            props: true,
            component: {
                template: `<furet-ui-space
                                v-bind:spaceId="spaceId"
                                v-bind:menuId="menuId"
                                v-bind:actionId="actionId"
                                v-bind:viewId="viewId"
                            />`,
                props: ['spaceId', 'menuId', 'actionId', 'viewId'],
                beforeRouteEnter: (to, from, next) => {
                    json_post_dispatch_all('/space/' + to.params.spaceId, to.params);
                    next();
                },
                beforeRouteUpdate: (to, from, next) => {
                    if (to.params.spaceId != from.params.spaceId) json_post_dispatch_all('/space/' + to.params.spaceId, to.params);
                    next();
                },
            },
            children: [
                {
                    path: 'menu/:menuId',
                    name: 'space_menu',
                    component: {
                        template: '<router-view></router-view>'
                    },
                    children: [
                        {
                            path: 'action/:actionId',
                            name: 'space_menu_action',
                            component: Action,
                            children: [
                                {
                                    path: 'view/:viewId',
                                    name: 'space_menu_action_view',
                                    props: true,
                                    component: View,
                                    children: [
                                        {
                                            path: 'data/:dataId/mode/:mode',
                                            name: 'space_menu_action_view_dataId',
                                            component: {template: '<div />'},  // never called
                                        },
                                    ],
                                },
                                {
                                    path: 'custom/view/:viewName',
                                    name: 'space_menu_action_custom_view',
                                    props: true,
                                    component: CustomeView,
                                },
                            ],
                        },
                    ],
                },
                {
                    path: 'action/:actionId',
                    name: 'space_action',
                    component: Action,
                    children: [
                        {
                            path: 'view/:viewId',
                            name: 'space_action_view',
                            props: true,
                            component: View,
                            children: [
                                {
                                    path: 'data/:dataId/mode/:mode',
                                    name: 'space_action_view_dataId',
                                    component: {template: '<div />'},  // never called
                                },
                            ],
                        },
                        {
                            path: 'custom/view/:viewName',
                            name: 'space_action_custom_view',
                            props: true,
                            component: CustomeView,
                        },
                    ],
                },
            ]
        },
    ],
});

export default router;
