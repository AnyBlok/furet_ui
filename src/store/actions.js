/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
export const UNITEST_CLEAR = ({ commit }) => {
    commit('CLEAR_CLIENT');
    commit('CLEAR_DATA');
    commit('CLEAR_GLOBAL');
    commit('CLEAR_LEFT_MENU');
    commit('CLEAR_RIGHT_MENU');
}
