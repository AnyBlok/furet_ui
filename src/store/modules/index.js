/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import global from './global';
import client from './client';
import rightmenu from './right-menu';
import leftmenu from './left-menu';
import data from './data';

export default {
    global,
    client,
    rightmenu,
    leftmenu,
    data,
}
