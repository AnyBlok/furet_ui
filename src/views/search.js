/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import Vue from 'vue';
import vSelect from 'vue-select'
Vue.component('v-select', vSelect)
import {json_post} from '../server-call';
import _ from 'underscore';

export const onChange = (obj, values) => {
    if (JSON.stringify(values) != obj.origin) {
        const filter = [];
        _.each(values, value => {
            const e = _.find(filter, f => f.key == value.key);
            if (e == undefined)
                filter.push({
                    key: value.key,
                    label: value.label,
                    value: value.value,
                    operator: value.operator,
                    type: value.type,
                })
            else {
                if (!Array.isArray(e.value)) e.value = [e.value];
                e.value.push(value.value)
                e.label = obj.labels[e.key] + ' : "' + e.value.join('", "') + '"';
            }
        });
        obj.origin = JSON.stringify(filter);
        obj.selected = filter;
        obj.$emit('updateFilter', filter);
    }
};

export const SearchBar = Vue.component('furet-ui-search-bar-view', {
    props: ['filter', 'search', 'model'],
    template: `
        <div class="field has-addons">
            <p class="control">
                <a class="button" v-on:click="onClick">
                    <i class="fa fa-search"></i>
                </a>
            </p>
            <p class="control">
                <v-select
                    multiple
                    v-bind:value.sync="selected"
                    v-bind:options="options"
                    v-bind:style="{backgroundColor: 'white'}"
                    v-bind:debounce="250"
                    v-bind:on-search="getOptions"
                    v-bind:on-change="onChange"
                />
            </p>
        </div>`,
    data () {
        return {
            selected: JSON.parse(JSON.stringify(this.filter)),
            origin: JSON.stringify(this.filter),
            options: [],
        }
    },
    computed: {
        labels () {
            const res = {};
            _.each(this.search, s => {res[s.key] = s.label});
            return res;
        }
    },
    methods: {
        getOptions (value, loading) {
            const self = this;
            loading(true);
            json_post('/data/search', {search: this.search, value, model: this.model}, {
                onSuccess (options) {self.options = options},
                onComplete () {loading(false)},
            });
        },
        onClick () {
            this.$emit('updateFilter', this.selected);
        },
        onChange(values) {
            onChange(this, values);
        },
    },
})
