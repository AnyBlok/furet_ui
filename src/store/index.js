/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import _ from 'underscore';
import Vue from 'vue';
import Vuex from 'vuex';
import * as actions from './actions';
import * as getters from './getters';
import modules from './modules';
import {i18n} from '../i18n';
import {router} from '../routes';
import moment from 'moment';

Vue.use(Vuex)
const debug = process.env.NODE_ENV !== 'production'

export const storeDef = {
    actions,
    getters,
    modules,
    strict: debug,
};

export const store = new Vuex.Store(storeDef);

export const dispatchAll = (datas) => {
    _.each(datas, data => {
        if (data.type) {
            switch (data.type) {
                case 'UPDATE_LOCALES':
                    _.each(data.locales, locale => {
                        i18n.setLocaleMessage(locale.locale, locale.messages)
                    });
                    break
                case 'SET_LOCALE':
                    i18n.locale = data.locale;
                    moment.locale(data.locale);
                    break
                case 'UPDATE_ROUTE':
                    if (data.name) router.push({name: data.name, params: data.params});
                    else if (data.path) router.push({path: data.path});
                    break;
                default:
                    const type = data.type;
                    delete data.type;
                    const process = data.process || 'commit';
                    store[process](type, data);
            }
        }
    });
}

export default store;
