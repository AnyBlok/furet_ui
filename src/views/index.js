/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import './clients';
import './search';
import './list';
import './thumbnail';
import './form';
import Vue from 'vue';
import plugin from '../plugin';

/**
 * Unknown view, use if no view found
 *
 * props:
 *  @viewName: Name of the custom view which is not available
 *  @viewType: Name of the standard view which is not available
 *
**/
export const ViewUnknown = Vue.component('furet-ui-view-unknown', {
    props: ['viewName'],
    template: `
        <div v-bind:style="{marginBottom: '20px'}">
            <section class="section box">
                <h1 class="title">{{$t('views.unknown.title', {name: viewName})}}</h1>
                <h2 class="subtitle">
                    {{$t('views.unknown.message')}}
                </h2>
            </section>
        </div>`,
});

plugin.set(['views'], {Unknown: {vue: 'furet-ui-view-unknown'}})


export const ViewIconUnknown = Vue.component('furet-ui-view-icon-unknown', {
    template: '<i class="fa fa-exclamation-triangle"></i>',
})

plugin.set(['views', 'icon'], {Unknown: 'furet-ui-view-icon-unknown'})
