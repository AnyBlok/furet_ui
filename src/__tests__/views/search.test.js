/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import chai from 'chai';
import Vue from 'vue';
import Buefy from 'buefy';
Vue.use(Buefy, {defaultIconPack: 'fa',});
import {store} from '../../store';
import {router} from '../../routes';
import {i18n} from '../../i18n';
import {SearchBar, onChange} from '../../views/search';

jest.mock('../../server-call')

describe('furet-ui-search-bar-view component', () => {
    const renderer = require('vue-server-renderer').createRenderer();
    beforeEach(() => {
        store.dispatch('UNITEST_CLEAR');
        router.push({path: '/'});
    });
    it('Render', () => {
        const vm = new Vue({
            el: document.createElement('div'),
            store,
            router,
            i18n,
            render: h => h(SearchBar, {props: {
                filter: [],
            }}),
        });
        renderer.renderToString(vm, (err, str) => {
            expect(str).toMatchSnapshot();
        });
    });
    it('onChange 1', () => {
        const obj = {
            origin: '',
            label: {},
            $emit (key, value) {
                expect(key).toBe('updateFilter'),
                chai.expect(value).to.deep.equal([]);
            },
        }
        onChange(obj, []);
        chai.expect(obj.selected).to.deep.equal([]);
        chai.expect(obj.origin).to.deep.equal('[]');
    });
    it('onChange 2', () => {
        const values = [
            {
                key: 'test',
                label: 'label',
                value: 'test',
                operator: 'equal',
                type: 'search',
            },
        ];
        const filter = [
            {
                key: 'test',
                label: 'label',
                value: 'test',
                operator: 'equal',
                type: 'search',
            },
        ]
        const obj = {
            origin: '',
            label: {},
            $emit (key, value) {
                expect(key).toBe('updateFilter'),
                chai.expect(value).to.deep.equal(filter);
            },
        }
        onChange(obj, values);
        chai.expect(obj.selected).to.deep.equal(filter);
        chai.expect(obj.origin).to.deep.equal(JSON.stringify(filter));
    });
    it('onChange 3', () => {
        const values = [
            {
                key: 'test',
                label: 'label',
                value: 'test 1',
                operator: 'equal',
                type: 'search',
            },
            {
                key: 'test',
                label: 'label',
                value: 'test 2',
                operator: 'equal',
                type: 'search',
            },
        ];
        const filter = [
            {
                key: 'test',
                label: 'label : "test 1", "test 2"',
                value: ['test 1', 'test 2'],
                operator: 'equal',
                type: 'search',
            },
        ]
        const obj = {
            origin: '',
            labels: {test: 'label'},
            $emit (key, value) {
                expect(key).toBe('updateFilter'),
                chai.expect(value).to.deep.equal(filter);
            },
        }
        onChange(obj, values);
        chai.expect(obj.selected).to.deep.equal(filter);
        chai.expect(obj.origin).to.deep.equal(JSON.stringify(filter));
    });
});
