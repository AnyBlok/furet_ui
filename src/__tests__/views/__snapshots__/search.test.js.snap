// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`furet-ui-search-bar-view component Render 1`] = `
<div data-server-rendered="true" class="field has-addons">
    <p class="control"><a class="button"><i class="fa fa-search"></i></a></p>
    <p class="control">
        <div debounce="250" class="dropdown v-select searchable" style="background-color:white;">
            <div class="dropdown-toggle"> <input type="search" value="" class="form-control" style="width:100%;"> <i role="presentation" class="open-indicator"></i>
                <div class="spinner" style="display:none;">Loading...</div>
            </div>
            <!---->
        </div>
    </p>
</div>
`;
